Components/Sensors/MTK3339.cs

```diff
@@ -29,95 +23,75 @@ public class MTK3339 : ISensor
-        /// <returns> Returns true if the GPS has a fix and false otherwise. </returns>
-        public bool Test() => HasFix();
-
-        /// <summary> Gets new readings from GPS. </summary>
-        public void UpdateState() => GetCoords();
-
-        /// <summary> Queries the GPS to see if it has a fix. </summary>
-        /// <returns> <c>true</c>, if fix was hased, <c>false</c> otherwise. </returns>
-        public bool HasFix()
+        private void Parse()
         {
-            WriteString(QUERY_GPGSA);
-            string[] Result = Read();
-            if (Result.Length < 3) { return false; }
-            return Result[2] != "1";
-        }
+            StringBuilder Builder = new StringBuilder();
 
-        public void HasFix(Action<bool> OnFinish)
-        {
-            if (!GettingHasFix)
+            while (true)
             {
-                Thread T = new Thread(() =>
+                char Data = BlockingGetChar();
+                Builder.Append(Data);
+                if (Data == '\n')
                 {
-                    GettingHasFix = true;
-                    OnFinish.Invoke(HasFix());
-                    GettingHasFix = false;
-                });
-                T.Start();
+                    if (Builder.ToString().Contains("$GPGGA"))
+                    {
+                        int? LatIndex = null;
+                        int? LngIndex = null;
+                        string[] SplitData = Builder.ToString().Split(',');
+                        for (int i = 0; i < SplitData.Length; i++)
+                        {
+                            switch (SplitData[i])
+                            {
+                                case "N": LatIndex = i - 1; break;
+                                case "S": LatIndex = -(i - 1); break;
+                                case "W": LngIndex = -(i - 1); break;
+                                case "E": LngIndex = i - 1; break;
+                            }
+                        }
+
+                        if (LatIndex.HasValue && LngIndex.HasValue)
+                        {
+                            Latitude = RawToDeg(SplitData[Math.Abs(LatIndex.Value)]) * Math.Sign(LatIndex.Value);
+                            Longitude = RawToDeg(SplitData[Math.Abs(LngIndex.Value)]) * Math.Sign(LngIndex.Value);
+                        }
+                    }
+                    Builder.Clear();
+                }
             }
         }
 
-        public void GetCoords(Action<Tuple<float, float>> OnFinish)
+        private char BlockingGetChar()
         {
-            if (!GettingCoords)
-            {
-                Thread T = new Thread(() =>
-                {
-                    GettingCoords = true;
-                    OnFinish.Invoke(GetCoords());
-                    GettingCoords = false;
-                });
-                T.Start();
-            }
+            byte[] NewChar = new byte[1];
+            while (UART.BytesAvailable() < 1) ;
```

Comment:
Hold on, won't this loop as fast as possible, consuming a full CPU core?

Reply:
The Linux scheduler will periodically come in and pause the process to run other stuff. I believe that by default the scheduler has a 10ms granularity, meaning it will pause the process in multiples of 10ms. 
However, adding a short delay may still be a good idea, as it will invoke the scheduler explicitly.